datajunction-server/datajunction_server/internal/nodes.py

```diff
@@ -480,24 +466,151 @@ def _update_node(
+):
+    """
+    Propagate the updated node's changes to all of its downstream children.
+    Some potential changes to the upstream node and their effects on downstream nodes:
+    - altered column names: may invalidate downstream nodes
+    - altered column types: may invalidate downstream nodes
+    - new columns: won't affect downstream nodes
+    """
+    processed = set()
+
+    # Each entry being processed is a list that represents the changelog of affected nodes
+    # The last entry in the list is the current node that's being processed
+    to_process = deque([[node.current, child] for child in node.children])
+
+    while to_process:
+        changelog = to_process.popleft()
+        child = changelog[-1]
+
+        # Only process if it hasn't already been processed before
+        if child.name not in processed:
+            processed.add(child.name)
+
+            node_validator = validate_node_data(
+                data=child,
+                session=session,
+            )
+
+            # Update the child with a new node revision if its columns or status have changed
+            if node_validator.differs_from(child):
+                new_revision = copy_existing_node_revision(child)
+                new_revision.version = str(
+                    Version.parse(child.version).next_major_version(),
+                )
+
+                new_revision.status = node_validator.status
+                new_revision = add_lineage_to_node(session, new_revision)
+
+                # Save which columns were modified and update the columns with the changes
+                updated_columns = node_validator.modified_columns(new_revision)
+                new_revision.columns = node_validator.columns
+
+                # Save the new revision of the child
+                new_revision.node = child.node
+                new_revision.node.current_version = new_revision.version
+                session.add(new_revision)
+                session.add(new_revision.node)
+
+                # Add grandchildren for processing
+                for grandchild in child.node.children:
+                    new_changelog = changelog + [grandchild]
+                    to_process.append(new_changelog)
+
+                # Record history event
+                history_events[child.name] = {
+                    "name": child.name,
+                    "current_version": new_revision.version,
+                    "previous_version": child.version,
+                    "updated_columns": sorted(list(updated_columns)),
+                }
+                event = History(
+                    entity_type=EntityType.NODE,
+                    entity_name=child.name,
+                    node=child.name,
+                    activity_type=ActivityType.STATUS_CHANGE,
+                    details={
+                        "upstreams": [
+                            history_events[entry.name]
+                            for entry in changelog
```

Comment:
I have a feeling this will cause the same nodes to be added to event history multiple times.

Reply:
That is the intention -- in the sense that we want all of the upstream actions up until this node to be recorded in the event history of that node, so that we know what happened up to the point this node got changed.

Reply:
I see! That makes sense...